BlockBuilder: Add the TSDBBuilder to build blocks and upload

codesome opened the pull request:
#### What this PR does

This PR adds the TSDBBuilder to build blocks and compact and upload. This is part of #8635; have a look at it for more details.

cc @narqo 

#### Checklist

- [X] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

pracucci reviewed (commented):
Nice work! I did a first pass and went very quickly through tests (I will get back to tests). The overall design LGTM. Since we're productionizing this component, I left several minor comments and few questions. Thanks!

narqo commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+		delete(b.tsdbs, tenant)
+
+		// Make a copy of userID because 'tenant' changes in the next iteration.
+		userID := tenant.id
```
> [..] So you don't have to worry about them anymore.

Since Go 1.23 is out now, I hope we will soon update `go.mod` Go 1.22, which fixed it on [language level][1]. After that we can wipe this piece of Go knowledge for better ;)

[1]: https://go.dev/wiki/LoopvarExperiment#my-test-fails-with-the-change-how-can-i-debug-it

narqo commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+					continue
+				}
+			}
+			// TODO(v): not all errors should terminate; see how it ingester handles them
```
We're addressing it separately https://github.com/grafana/mimir/pull/9006 👍

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+// where the sample was not put in the TSDB because it was discarded or was already processed before.
+// lastEnd: "end" time of the previous block building cycle.
+// currEnd: end time of the block we are looking at right now.
+func (b *TSDBBuilder) Process(ctx context.Context, rec *kgo.Record, lastBlockMax, blockMax int64, recordProcessedBefore bool) (_ bool, err error) {
```
At the very moment it is not called concurrently. But I think we will soon introduce parallel processing, in which case we will need the locking. So I would keep the locking around for now.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+
+	db, err := tsdb.Open(udir, userLogger, nil, &tsdb.Options{
+		RetentionDuration:           0,
+		MinBlockDuration:            2 * time.Hour.Milliseconds(),
+		MaxBlockDuration:            2 * time.Hour.Milliseconds(),
```
The 1h is something controlled by the overarching block builder using config and processing 1h data appropriately from Kafka. The TSDB still needs to work with 2h block boundaries like rest of the Mimir in case block builder tries to process some samples before the 1h block we are looking at.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+		userID: userID,
+	}
+
+	db, err := tsdb.Open(udir, userLogger, nil, &tsdb.Options{
```
Added to TODO!

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+	return nil
+}
+
+func (u *userTSDB) PreCreation(_ labels.Labels) error                     { return nil }
+func (u *userTSDB) PostCreation(_ labels.Labels)                          {}
+func (u *userTSDB) PostDeletion(_ map[chunks.HeadSeriesRef]labels.Labels) {}
```
It's still to be built. In general, all limits are still to be built because all of them have a fundamental problem of how to apply when they change, i.e. ingesters will apply the current limits to current data, but block builder will be applying current limits to past data, so its application will differ from ingesters if the config was changed.

pracucci replied:
Ok this is quite concerning to me. I would like to see this addressed asap.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+	return udb, nil
+}
+
+func (b *TSDBBuilder) tsdbDir(tenant tsdbTenant) string {
```
Will remove it for now. I think it is the only place.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+}
+
+func (u *userTSDB) compactEverything(ctx context.Context) error {
+	blockRange := 2 * time.Hour.Milliseconds()
```
I think the answer to the other comment applies here as well: https://github.com/grafana/mimir/pull/8978#discussion_r1717489246

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+			return numBlocks, err
+		}
+
+		delete(b.tsdbs, tenant)
```
If there is some error in between for some tenants, we dont clear the map in that case. I don't have strong opinion on what to do in case of such errors, but I think we should not discard all the TSDBs if they were not uploaded. I moved this delete() to happen after we have uploaded the blocks for this tenant.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+
+	// Clear the map so that it can be released from the memory. Not setting to nil in case
+	// we want to reuse the TSDBBuilder.
+	b.tsdbs = make(map[tsdbTenant]*userTSDB)
```
We do not reset the map in case of compaction failure. The DBs that are not compacted and uploaded yet are still in the map.

codesome replied:
Actually, we are clearing the map even if some tenants failed to upload. I will make it such that we reset the map only if it was empty. This plays well with the delete() that we have on the map (commented about delete() above).

pracucci replied:
Ok. It's still a bit unclear to me what happens when an error happens. Let's say we haven't been to successfully compact and upload some blocks... what's next? What do we do?

pracucci replied:
Let's not block on this discussion, but we can continue this conversation on Slack (please kick it off).

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+
+// CompactAndUpload compacts the blocks of all the TSDBs and uploads them.
+// blockUploaderForUser should give a function to upload the blocks of a particular user.
+func (b *TSDBBuilder) CompactAndUpload(ctx context.Context, blockUploaderForUser func(context.Context, string) blockUploader) (int, error) {
```
You are right, its all a small wrapper. I can have a simple function to upload the blocks with arguments `(tenantID, dbDir, []blockIDs)`

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+	}
+	numBlocks := 0
+	for tenant, db := range b.tsdbs {
+		if err := db.compactEverything(ctx); err != nil {
```
It was not intentional. I guess some changes got rolled back. We had compaction in parallel. I am thinking to re-use ShipConcurrency config for the compaction concurrency for now.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+		if err := db.Close(); err != nil {
+			return err
+		}
+		delete(b.tsdbs, userID)
```
Inside this function I think we dont need the delete() I have removed it and made sure we close all DBs even if any of them error out and return the first error, and reset the map regardless.

codesome commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,398 @@
+	return numBlocks, err
+}
+
+func (b *TSDBBuilder) Close() error {
```
Added a comment
```
// Close closes all DBs and deletes their data directories.
// This functions is useful when block builder has faced some unrecoverable error
// and has to discard the block building for the current cycle.
```

codesome commented:
I have addressed most of the comments and answered the questions. Few pending things:
* Unit test for a edge case (https://github.com/grafana/mimir/pull/8978#discussion_r1716542188). I will do it in this PR.
* Handling soft errors on append: being handled in https://github.com/grafana/mimir/pull/9006, so we can skip it in this PR
* Handling limits: it is better to be done in a follow up PR
* Suggested optimisation about 1 sample per series in most cases (https://github.com/grafana/mimir/pull/8978#discussion_r1716520369). I can do it in this PR. But if reviews for other parts are done before I get to it, we can do it in a follow up PR.

pracucci commented on pkg/blockbuilder/tsdb.go:
```diff
@@ -0,0 +1,393 @@
+				return err
+			}
+
+			delete(b.tsdbs, tenant)
```
I just realised there's a race here: `t.tsdbs` is manipulated in `eg.Go()` which means there could be multiple goroutines manipulating it.

codesome replied:
Race should be fixed now

pracucci reviewed (approved):
Thanks for addressing my comments. I'm fine merging this PR as is (modulo fixing the race condition I just found out) and then you can open follow up PRs for the remaining comments. I would like to kick off the work on limits as soon as possible (I totally missed it wasn't done in the prototype).
